src/nvim/buffer.c

```diff
@@ -3107,42 +3222,54 @@ build_stl_str_hl (
       else
         str = path_tail(NameBuff);
       break;
-
+    }
     case STL_VIM_EXPR:     /* '{' */
-      itemisflag = TRUE;
-      t = p;
-      while (*s != '}' && *s != NUL && p + 1 < out + outlen)
-        *p++ = *s++;
-      if (*s != '}')            /* missing '}' or out of space */
+    {
+      itemisflag = true;
+
+      // Attempt to copy the expression to evaluate into
+      // the output buffer as a null-terminated string.
+      char_u *t = out_p;
+      while (*fmt_p != '}' && *fmt_p != NUL && out_p < out_end_p)
+        *out_p++ = *fmt_p++;
+      if (*fmt_p != '}')            /* missing '}' or out of space */
         break;
-      s++;
-      *p = 0;
-      p = t;
+      fmt_p++;
+      *out_p = 0;
+
+      // Move our position in the output buffer
+      // to the beginning of the expression
+      out_p = t;
+
+      // { Evaluate the expression
 
+      // Store the current buffer number as a string variable
       vim_snprintf((char *)tmp, sizeof(tmp), "%d", curbuf->b_fnum);
       set_internal_string_var((char_u *)"actual_curbuf", tmp);
 
-      o_curbuf = curbuf;
-      o_curwin = curwin;
-      curwin = wp;
-      curbuf = wp->w_buffer;
+      // Evaluate the expression using the correct window
+      WITH_WINDOW(wp, {
+        // Note: The result stored in `t` is unused.
+        str = eval_to_string_safe(out_p, &t, use_sandbox);
+      });
```

Comment:
Sadly, this is not equivalent. `switch_win()` calls `block_autocmds()` which I am assuming is the cause of some weird behavior with this PR. Test case:

```vim
    autocmd! TermOpen * echom 'termopen' localtime()
    function! Foo()
      doautocmd TermOpen
    endfunction
    set statusline=%{Foo()}
```

The autocmd is not triggered by statusline evaluation ("No matching autocmds" is reported every time statusline redraws).

Perhaps `WITH_WINDOW` could take a parameter which prevents disabling of autocmds?